### PR TITLE
FIX #3123 : added csvSeparator property in table widget for csv download

### LINKED ISSUE #3123: [Feature] Provide config options for CSV downloads in table
1. Need to specify csv separator and number separator in table download
2. In France ; is used instead of , as CSV separators.
3. , is used as number separator

### PR DESCRIPTION
**Pull request by vibhandikyash**
## Description

> Added Download Config for downloading CSV data of the table. Added option for separator configuration.

Fixes #3123 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

> Tested locally.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: FEATURE/3123-config-option-for-csv-download 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 54.89 **(0.02)** | 37 **(0.04)** | 33.21 **(0.02)** | 55.47 **(0.02)**
 :red_circle: | app/client/src/components/designSystems/appsmith/TableComponent/index.tsx | 27.27 **(0)** | 2.6 **(-0.03)** | 15.79 **(0)** | 25.68 **(0)**
 :green_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 72.33 **(0.15)** | 64.46 **(-0.06)** | 67.03 **(0)** | 71.3 **(0.15)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 50.71 **(0.24)** | 38.6 **(0.88)** | 36.21 **(0)** | 55.35 **(0.27)**
 :green_circle: | app/client/src/utils/migrations/TableWidget.ts | 75.35 **(1.87)** | 70.45 **(1.28)** | 83.33 **(2.08)** | 78.91 **(1.6)**
 :green_circle: | app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts | 29.82 **(0.62)** | 8.43 **(0)** | 47.37 **(2.93)** | 31.19 **(0.63)**
 :green_circle: | app/client/src/widgets/TableWidget/index.tsx | 27.59 **(0.18)** | 6.29 **(0)** | 25.49 **(0)** | 27.78 **(0.2)**</details>

### CODE REVIEW

File: app/client/src/entities/Widget/utils.test.ts [hunk 1]
```diff
@@ -166,6 +167,12 @@ describe("getAllPathsFromPropertyConfig", () => {
         defaultSearchText: {
           type: "TEXT",
         },
+        csvSeparator: {
+          type: "TEXT",
+          params: {
+            allowedValues: [",", ";"],
```
Comment (SatishGandham) [inline app/client/src/entities/Widget/utils.test.ts:173]: Why are we restricting on the values they can use?
If we are for some reason, pipe (|) is also commonly used.

File: app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx [hunk 2]
```diff
@@ -77,6 +77,7 @@ interface TableProps {
   isVisibleFilters?: boolean;
   isVisiblePagination?: boolean;
   isVisibleSearch?: boolean;
+  csvSeparator: string;
```
Comment (SatishGandham) [inline app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx:80]: csv is comma separated values. Can we have a more accurate name?
CC: @somangshu

**Review by SatishGandham — CHANGES_REQUESTED** [review]

File: app/client/src/entities/Widget/utils.test.ts (hunk 1 above)
Comment (somangshu) [reply app/client/src/entities/Widget/utils.test.ts:173]: @vibhandikyash please explain this

File: app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx (hunk 2 above)
Comment (somangshu) [reply app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx:80]: Maybe `csvDelimiter`?

Comment (vibhandikyash) [reply app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx:80]: +1 for @somangshu Maybe we can go with this one. @SatishGandham kindly confirm.

File: app/client/src/entities/Widget/utils.test.ts (hunk 1 above)
Comment (vibhandikyash) [reply app/client/src/entities/Widget/utils.test.ts:173]: @SatishGandham Usually for CSV internal formats are only two of them comma (,) and semicolon(;). If you want we can remove the validation.

Comment (somangshu) [reply app/client/src/entities/Widget/utils.test.ts:173]: Rather, can we include the Pipe ( | ) as well? @vibhandikyash

File: app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx (hunk 2 above)
Comment (SatishGandham) [reply app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx:80]: How about just delimiter, if that doesn't work we can go with csvDelimiter

File: app/client/src/entities/Widget/utils.test.ts (hunk 1 above)
Comment (SatishGandham) [reply app/client/src/entities/Widget/utils.test.ts:173]: I think we should remove the limitation. Sometime they may use || or completely different delimiter when , ; or | is part of the data.

File: app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx (hunk 2 above)
Comment (vibhandikyash) [reply app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx:80]: @somangshu Thoughts?

File: app/client/src/entities/Widget/utils.test.ts (hunk 1 above)
Comment (vibhandikyash) [reply app/client/src/entities/Widget/utils.test.ts:173]: okay @SatishGandham, Noted. Will remove the validation completely.
cc: @somangshu

File: app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx (hunk 2 above)
Comment (somangshu) [reply app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx:80]: @vibhandikyash going with ☝️

**Review by somangshu — DISMISSED** [review]
LGTM

**Comment by somangshu** [comment]
From the deploy preview:
- The helper text for the new field needs to change to > The character used for separating the CSV download file. 
- For older widgets the evaluated value does not show up, Neither does the download format right. **(Migrations issues)**
- Merge the CSV separator field to the header option section and get rid of the download config section

@vibhandikyash please resolve the above

**Comment by somangshu** [comment]
Updated the above comment @vibhandikyash

File: app/client/src/utils/migrations/TableWidget.ts [hunk 3]
```diff
@@ -243,6 +243,7 @@ export const migrateTableWidgetHeaderVisibilityProperties = (
         child.isVisibleDownload = true;
         child.isVisibleCompactMode = true;
         child.isVisiblePagination = true;
+        child.delimeter = ",";
```
Comment (somangshu) [inline app/client/src/utils/migrations/TableWidget.ts:246]: Do we also need to upgrade the version?

**Comment by somangshu** [comment]
@vibhandikyash the jest tests have failed here because the table widget property do not match, Please fix this.

**Comment by somangshu** [comment]
On the deploy preview I still see the migration issue, The default value for the property is not evaluated
Screenshot 2021-08-13 at 12 57 52 PM

**Review by somangshu — APPROVED** [review]
Approving for re-running the tests

**Comment by somangshu** [comment]
@vibhandikyash the migration issue on DP still exists, Check [this](https://deploy-preview-6291--appsmith-editor-ce.netlify.app/applications/60aba4455123fa30d6c393f7/pages/60aba4455123fa30d6c393f9/edit) app

**Review by somangshu — APPROVED** [review]

**Comment by somangshu** [comment]
> @vibhandikyash the migration issue on DP still exists, Check [this](https://deploy-preview-6291--appsmith-editor-ce.netlify.app/applications/60aba4455123fa30d6c393f7/pages/60aba4455123fa30d6c393f9/edit) app

This works fine, Migration works correctly

File: app/client/src/entities/Widget/utils.test.ts [hunk 4]
```diff
@@ -120,6 +121,7 @@ describe("getAllPathsFromPropertyConfig", () => {
         defaultSearchText: EvaluationSubstitutionType.TEMPLATE,
         defaultSelectedRow: EvaluationSubstitutionType.TEMPLATE,
         isVisible: EvaluationSubstitutionType.TEMPLATE,
+        delimeter: EvaluationSubstitutionType.TEMPLATE,
```
Comment (SatishGandham) [inline app/client/src/entities/Widget/utils.test.ts:124]: Its delimiter, not delimeter

**Review by SatishGandham — CHANGES_REQUESTED** [review]

**Review by SatishGandham — APPROVED** [review]

**Review by somangshu — APPROVED** [review]

**Pull request merged**